**Introduction- **

Price of a property is one of the most important decision criterion when people buy homes. Real state firms need to be consistent in their pricing in order to attract buyers . Having a predictive model for the same will be great tool to have , which in turn can also be used to tweak development of properties , putting more emphasis on qualities which increase the value of the property.

Problem Statement- 

We have been given two datasets , housing_train.csv and housing_test.csv . We need to use data housing_train to build predictive model for response variable “Price”. Housing_test data contains all other factors except “Price”, we need to predict that using the model that we developed.

**Loading the data (Train & test both)- **

In [33]:
h_train=read.csv("../input/real-estate-data-sets/housing_train.csv",stringsAsFactors = F)
h_test=read.csv("../input/house-data-sets/house_test.csv",stringsAsFactors = F)

**Data Dictionary- **

Each row represnts charactaristic of a single property . Many categorical data has been coded to mask the data

Suburb : categorical :: Which subsurb the property is located in

Address : categorical :: short address

Rooms : numeric :: Number of Rooms

Type : categorical :: type of the property

Price : numeric :: This is the target variable, price of the property

Method : categorical :: method for selling

SellerG : categorical :: Name of the seller

Distance : numeric :: distance from the city center Postcode : categorical :: postcode of the property

Bedroom2 : Numeric :: numbers of secondary bedrooms (this is different from rooms)

Bathroom : numeric :: number of bathrooms

Car : numeric :: number of parking spaces

Landsize : numeric :: landsize

BuildingArea : numeric :: buildup area

YearBuilt : numeric :: year of building

CouncilArea : numeric :: council area to which the propery belongs

**Glimpse of our Data**

Let us see first few records of our train & test data using ‘head’ function as shown below. It is evident that our test data contains every fields except response variable ‘Price’.

In [34]:
head(h_train)

Suburb        Address        Rooms Type Price   Method SellerG Distance
1 Brunswick     52 Evans St    3     h    1650000 S      Nelson   5.2    
2 Reservoir     85 Radford Rd  5     h     791000 S      Ray     11.2    
3 Newport       99 Anderson St 3     h     785000 S      RT       8.4    
4 Brighton East 4/377 South Rd 2     u     755000 SP     Buxton  10.7    
5 Hawthorn East 3 Jaques St    5     h    2500000 VB     RT       7.5    
6 Hawthorn East 75 Leura Gr    3     h    3020000 S      Hooper   7.5    
  Postcode Bedroom2 Bathroom Car Landsize BuildingArea YearBuilt CouncilArea
1 3056      3        1        2  495      141          1920      Moreland   
2 3073      4        3        1  961       NA            NA      Darebin    
3 3015      3        1        1  185       NA            NA      Hobsons Bay
4 3187     NA       NA       NA   NA       NA            NA                 
5 3123      5        3        3  757      240          1925      Boroondara 
6 3123      3        2        2  832       NA            NA      Boroondara

In [35]:
head(h_test)

Suburb     Address             Rooms Type Method SellerG      Distance
1 Abbotsford 6/241 Nicholson St  1     u    S      Biggin       2.5     
2 Abbotsford 403/609 Victoria St 2     u    S      Dingle       2.5     
3 Abbotsford 106/119 Turner St   1     u    SP     Purplebricks 2.5     
4 Abbotsford 22 Park St          4     h    S      Biggin       2.5     
5 Abbotsford 78 Yarra St         3     h    S      LITTLE       2.5     
6 Abbotsford 13/11 Nicholson St  3     t    S      Beller       2.5     
  Postcode Bedroom2 Bathroom Car Landsize BuildingArea YearBuilt CouncilArea
1 3067      1        1        1    0       NA            NA      Yarra      
2 3067     NA       NA       NA   NA       NA            NA                 
3 3067     NA       NA       NA   NA       NA            NA                 
4 3067     NA       NA       NA   NA       NA            NA                 
5 3067      2        1        1  138      105          1890      Yarra      
6 3067      3        2        2    0       NA          2010      Yarra

In order to glimpse data & its data type we need to use function glimpse from the package ‘dplyr’.

In [36]:
library(dplyr)

In [37]:
glimpse(h_train)

Observations: 7,536
Variables: 16
$ Suburb       <chr> "Brunswick", "Reservoir", "Newport", "Brighton East", "H…
$ Address      <chr> "52 Evans St", "85 Radford Rd", "99 Anderson St", "4/377…
$ Rooms        <int> 3, 5, 3, 2, 5, 3, 3, 3, 4, 2, 3, 2, 2, 2, 4, 3, 2, 2, 4,…
$ Type         <chr> "h", "h", "h", "u", "h", "h", "h", "h", "h", "u", "t", "…
$ Price        <int> 1650000, 791000, 785000, 755000, 2500000, 3020000, 78000…
$ Method       <chr> "S", "S", "S", "SP", "VB", "S", "VB", "VB", "PI", "S", "…
$ SellerG      <chr> "Nelson", "Ray", "RT", "Buxton", "RT", "Hooper", "hockin…
$ Distance     <dbl> 5.2, 11.2, 8.4, 10.7, 7.5, 7.5, 13.9, 11.2, 12.8, 11.8, …
$ Postcode     <int> 3056, 3073, 3015, 3187, 3123, 3123, 3165, 3127, 3033, 31…
$ Bedroom2     <int> 3, 4, 3, NA, 5, 3, 3, 3, NA, 2, NA, 2, 2, 2, 4, NA, 2, 2…
$ Bathroom     <int> 1, 3, 1, NA, 3, 2, 1, 2, NA, 2, NA, 1, 2, 1, 2, NA, 1, 1…
$ Car          <int> 2, 1, 1, NA, 3, 2, 1, 4, NA, 2, NA, 1, 1, 1, 1, NA, 1, 4…
$ Landsize     <in

In [38]:
glimpse(h_test)

Observations: 1,885
Variables: 15
$ Suburb       <chr> "Abbotsford", "Abbotsford", "Abbotsford", "Abbotsford", …
$ Address      <chr> "6/241 Nicholson St", "403/609 Victoria St", "106/119 Tu…
$ Rooms        <int> 1, 2, 1, 4, 3, 3, 3, 1, 1, 2, 3, 1, 3, 2, 3, 3, 2, 3, 3,…
$ Type         <chr> "u", "u", "u", "h", "h", "t", "u", "u", "u", "h", "t", "…
$ Method       <chr> "S", "S", "SP", "S", "S", "S", "S", "S", "SP", "S", "SP"…
$ SellerG      <chr> "Biggin", "Dingle", "Purplebricks", "Biggin", "LITTLE", …
$ Distance     <dbl> 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2.5, 2…
$ Postcode     <int> 3067, 3067, 3067, 3067, 3067, 3067, 3067, 3067, 3067, 30…
$ Bedroom2     <int> 1, NA, NA, NA, 2, 3, 3, NA, 1, 2, NA, 1, 3, NA, 3, 2, 2,…
$ Bathroom     <int> 1, NA, NA, NA, 1, 2, 2, NA, 1, 2, NA, 1, 2, NA, 2, 1, 1,…
$ Car          <int> 1, NA, NA, NA, 1, 2, 2, NA, 1, 2, NA, 1, 1, NA, 2, 1, 1,…
$ Landsize     <int> 0, NA, NA, NA, 138, 0, 4290, NA, 0, 98, NA, 0, 120, NA, …
$ BuildingArea <in

Train data contain 7536 observations or rows & 16 variables whereas test data has 1885 observations & 15 variables. 

**Data Preparation**

We’ll combine our two datasets so that we do not need to prepare data separately for them. However before combining them, we’ll need to add response column to test because number of columns need to be same for two datasets to stack vertically. Also we will add an identifier column to both train & test data sets so that we can separate them after our data preparation.

In [39]:
h_test$Price=NA

h_train$data="train"
h_test$data="test"

h=rbind(h_train,h_test)

Let us now glimpse our combined data sets ‘h’.

In [40]:
glimpse(h)

Observations: 9,421
Variables: 17
$ Suburb       <chr> "Brunswick", "Reservoir", "Newport", "Brighton East", "H…
$ Address      <chr> "52 Evans St", "85 Radford Rd", "99 Anderson St", "4/377…
$ Rooms        <int> 3, 5, 3, 2, 5, 3, 3, 3, 4, 2, 3, 2, 2, 2, 4, 3, 2, 2, 4,…
$ Type         <chr> "h", "h", "h", "u", "h", "h", "h", "h", "h", "u", "t", "…
$ Price        <int> 1650000, 791000, 785000, 755000, 2500000, 3020000, 78000…
$ Method       <chr> "S", "S", "S", "SP", "VB", "S", "VB", "VB", "PI", "S", "…
$ SellerG      <chr> "Nelson", "Ray", "RT", "Buxton", "RT", "Hooper", "hockin…
$ Distance     <dbl> 5.2, 11.2, 8.4, 10.7, 7.5, 7.5, 13.9, 11.2, 12.8, 11.8, …
$ Postcode     <int> 3056, 3073, 3015, 3187, 3123, 3123, 3165, 3127, 3033, 31…
$ Bedroom2     <int> 3, 4, 3, NA, 5, 3, 3, 3, NA, 2, NA, 2, 2, 2, 4, NA, 2, 2…
$ Bathroom     <int> 1, 3, 1, NA, 3, 2, 1, 2, NA, 2, NA, 1, 2, 1, 2, NA, 1, 1…
$ Car          <int> 2, 1, 1, NA, 3, 2, 1, 4, NA, 2, NA, 1, 1, 1, 1, NA, 1, 4…
$ Landsize     <in

From above I can see many missing values or NAs in my data. Let us see how many NAs are there for each variables. We will use ‘lapply’ function to arrive at those NA values who in list format as shown below.

In [41]:
lapply(h,function(x) sum(is.na(x)))

$Suburb
[1] 0

$Address
[1] 0

$Rooms
[1] 0

$Type
[1] 0

$Price
[1] 1885

$Method
[1] 0

$SellerG
[1] 0

$Distance
[1] 0

$Postcode
[1] 0

$Bedroom2
[1] 1978

$Bathroom
[1] 1978

$Car
[1] 1978

$Landsize
[1] 1985

$BuildingArea
[1] 5269

$YearBuilt
[1] 4660

$CouncilArea
[1] 0

$data
[1] 0

From above we can see that we have missing values in columns like Price(1885), Bedroom2 (1978), Bathroom (1978), Car(1978), Landsize(1985), BuildingArea (5269) & YearBuilt(4660).

We know that we purposely add Price column in our test data set & add NAs to that column. We are not going to impute for Price column & will go ahead imputing  NAs for remaining columns with central tendency of train data like mean or median. We will be doing this later as we progress.

**Transforming character variables into Dummy Variables-** 

Since categorical variables are known to hide and mask lots of interesting information in a data set.So It’s crucial we’d hunt down those categorical variables in the data set, and dig out as much information as you can. One  way is to convert the same into dummy variables.

Let us analyse our first categorical variable ‘Suburb’, the frequency table of which gives the following result. Basically its is nothing but categorical tabulation of data with the variable and its frequency as shown below. For example there are 52 observations which has Abbotsford as Suburb.

In [43]:
table(h$Suburb)


        Abbotsford         Aberfeldie       Airport West        Albert Park 
                52                 29                 65                 44 
            Albion         Alphington             Altona       Altona North 
                26                 29                 51                 71 
          Armadale         Ascot Vale          Ashburton            Ashwood 
                71                103                 62                 45 
  Avondale Heights          Balaclava             Balwyn       Balwyn North 
                69                 18                112                135 
         Bellfield          Bentleigh     Bentleigh East           Box Hill 
                13                127                241                 44 
         Braybrook           Brighton      Brighton East           Brooklyn 
                38                150                130                 15 
         Brunswick     Brunswick East     Brunswick West            Bulleen

**Writing a Function to create Dummy-** 

We will write a function which will be applied to all categorical variables so as to convert them into dummies.

In [45]:
CreateDummies=function(data,var,freq_cutoff=0){
t=table(data[,var])
t=t[t>freq_cutoff]
t=sort(t)
categories=names(t)[-1]
for( cat in categories){
name=paste(var,cat,sep="_")
name=gsub(" ","",name)
name=gsub("-","_",name)
name=gsub("\\?","Q",name)
name=gsub("<","LT_",name)
name=gsub("\\+","",name)
name=gsub("\\/","_",name)
name=gsub(">","GT_",name)
name=gsub("=","EQ_",name)
name=gsub(",","",name)
data[,name]=as.numeric(data[,var]==cat)
}
data[,var]=NULL
return(data)
}

Let me explain the function ‘CreateDummies’ we just created

t=table(data[,var]) this bit creates a frequency table for the given categorical column. t here is now simply a table which contains names as categories of the categorical variable and their frequency in the data.

t=t[t>freq_cutoff] this line of code removes those categories from the table which have frequencies below the frequency cutoff. ( this is a subjective choice)

‘t=sort(t)’ this line simple sorts the remaining table in ascending order

categories=names(t)[-1] since we sorted the table in ascending manner in the previous line, first category here has least count. In this line of code we are taking out the category names except the first one ( which has least count), thus making n-1 dummies from the remaining categories.

name=paste(var,cat,sep=“_“) all the dummy vars that we intend to make, need to have some name. this line of code creates that name by concatenating variable name with category name with an _.

name=gsub(" “,”“,name) subsequent lines like these using gsub are essentially cleaning up the name. Since we dont have any control over what the categories can be, we are removing special characters and spaces in the code in an automated fashion.

data[,name]=as.numeric(data[,var]==cat) once we have a cleaned up name, this line creates the dummy var for that particular category.

data[,var]=NULL once we are done creating dummies for the variable using for loop. Variable is removed from the data in this line.


Let us have a look at our categorical variables by writing following lines of codes

In [46]:
names(h)[sapply(h,function(x) is.character(x))]

[1] "Suburb"      "Address"     "Type"        "Method"      "SellerG"    
[6] "CouncilArea" "data"

Now we will check for High-Cardinality in the categorical variables i.e we will check for variables with many distinct values. We will discard those variables from our modelling. Because including these attributes by standard dummy encoding increases the dimensionality of the data to such an extent that either the classification technique is unable to process them or if one would use some regularized linear technique that is able to cope with huge dimensions, it leads to a model with thousands or even millions of features, thereby losing the often required comprehensibility aspect.

**Checking High-Cardinality in our categorical variables- **

In [47]:
length(unique(h$Suburb))

[1] 142

In [48]:
length(unique(h$Address))

[1] 9324

In [49]:
length(unique(h$Type))

[1] 3

In [50]:
length(unique(h$SellerG))

[1] 198

In [51]:
length(unique(h$CouncilArea))

[1] 20

We will ignore variable ‘Address’ for their High-Cardinality as shown above. Further we will ignore variable data for obvious reason.

In [52]:
h=h %>% select(-Address)

Let us make dummies for rest of the variables by using for loop.

In [53]:
cat_cols=c("Suburb","Type","Method","SellerG","CouncilArea")

for (cat in cat_cols){
  h=CreateDummies(h,cat,100)
}

Now This dummy creation has increased number of variables to 85 as we can glimpse the same as shown below.

In [54]:
glimpse(h)

Observations: 9,421
Variables: 85
$ Rooms                    <int> 3, 5, 3, 2, 5, 3, 3, 3, 4, 2, 3, 2, 2, 2, 4,…
$ Price                    <int> 1650000, 791000, 785000, 755000, 2500000, 30…
$ Distance                 <dbl> 5.2, 11.2, 8.4, 10.7, 7.5, 7.5, 13.9, 11.2, …
$ Postcode                 <int> 3056, 3073, 3015, 3187, 3123, 3123, 3165, 31…
$ Bedroom2                 <int> 3, 4, 3, NA, 5, 3, 3, 3, NA, 2, NA, 2, 2, 2,…
$ Bathroom                 <int> 1, 3, 1, NA, 3, 2, 1, 2, NA, 2, NA, 1, 2, 1,…
$ Car                      <int> 2, 1, 1, NA, 3, 2, 1, 4, NA, 2, NA, 1, 1, 1,…
$ Landsize                 <int> 495, 961, 185, NA, 757, 832, 710, 816, NA, 0…
$ BuildingArea             <int> 141, NA, NA, NA, 240, NA, NA, NA, NA, 80, NA…
$ YearBuilt                <int> 1920, NA, NA, NA, 1925, NA, 1966, NA, NA, 20…
$ data                     <chr> "train", "train", "train", "train", "train",…
$ Suburb_Doncaster         <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
$ Suburb_AscotVale

**Imputing Missing Values (NAs)- **

We have seen earlier that our data has missing values & we need to impute it with mean of train data.Also we will exclude such imputations in variables data & Price. Shared below For loop demonstrate the same.

In [55]:
for(col in names(h)){
if(sum(is.na(h[,col]))>0 & !(col %in% c("data","Price"))){
h[is.na(h[,col]),col]=mean(h[h$data=='train',col],na.rm=T)
}
}

Let us see if there are still any NAs in our data using lapply function again.

In [56]:
lapply(h,function(x) sum(is.na(x)))

$Rooms
[1] 0

$Price
[1] 1885

$Distance
[1] 0

$Postcode
[1] 0

$Bedroom2
[1] 0

$Bathroom
[1] 0

$Car
[1] 0

$Landsize
[1] 0

$BuildingArea
[1] 0

$YearBuilt
[1] 0

$data
[1] 0

$Suburb_Doncaster
[1] 0

$Suburb_AscotVale
[1] 0

$Suburb_Footscray
[1] 0

$Suburb_MooneePonds
[1] 0

$Suburb_Thornbury
[1] 0

$Suburb_Hampton
[1] 0

$Suburb_Yarraville
[1] 0

$Suburb_Balwyn
[1] 0

$Suburb_MalvernEast
[1] 0

$Suburb_Camberwell
[1] 0

$Suburb_Carnegie
[1] 0

$Suburb_PortMelbourne
[1] 0

$Suburb_Bentleigh
[1] 0

$Suburb_PascoeVale
[1] 0

$Suburb_BrightonEast
[1] 0

$Suburb_Hawthorn
[1] 0

$Suburb_BalwynNorth
[1] 0

$Suburb_Coburg
[1] 0

$Suburb_Northcote
[1] 0

$Suburb_Kew
[1] 0

$Suburb_Brighton
[1] 0

$Suburb_Glenroy
[1] 0

$Suburb_GlenIris
[1] 0

$Suburb_Essendon
[1] 0

$Suburb_Brunswick
[1] 0

$Suburb_SouthYarra
[1] 0

$Suburb_StKilda
[1] 0

$Suburb_Preston
[1] 0

$Suburb_Richmond
[1] 0

$Suburb_BentleighEast
[1] 0

$Suburb_Reservoir
[1] 0

$Type_u
[1] 0

$Type_h
[1] 0

$Method_SP
[1] 0

$Method_PI
[1] 0

$Method_S
[1] 0

$SellerG_Kay
[1] 0

$SellerG_Hodges
[1] 0

$SellerG_McGrath
[1] 0

$SellerG_Noel
[1] 0

$SellerG_Gary
[1] 0

$SellerG_Jas
[1] 0

$SellerG_Miles
[1] 0

$SellerG_Greg
[1] 0

$SellerG_Sweeney
[1] 0

$SellerG_RT
[1] 0

$SellerG_Fletchers
[1] 0

$SellerG_Woodards
[1] 0

$SellerG_Brad
[1] 0

$SellerG_Biggin
[1] 0

$SellerG_Ray
[1] 0

$SellerG_Buxton
[1] 0

$SellerG_Marshall
[1] 0

$SellerG_Barry
[1] 0

$SellerG_hockingstuart
[1] 0

$SellerG_Jellis
[1] 0

$SellerG_Nelson
[1] 0

$CouncilArea_Whitehorse
[1] 0

$CouncilArea_Manningham
[1] 0

$CouncilArea_Brimbank
[1] 0

$CouncilArea_HobsonsBay
[1] 0

$CouncilArea_Bayside
[1] 0

$CouncilArea_Melbourne
[1] 0

$CouncilArea_Banyule
[1] 0

$CouncilArea_PortPhillip
[1] 0

$CouncilArea_Yarra
[1] 0

$CouncilArea_Maribyrnong
[1] 0

$CouncilArea_Stonnington
[1] 0

$CouncilArea_GlenEira
[1] 0

$CouncilArea_Darebin
[1] 0

$CouncilArea_MooneeValley
[1] 0

$CouncilArea_Moreland
[1] 0

$CouncilArea_Boroondara
[1] 0

$CouncilArea_
[1] 0

**Data Separation-**

Now we are done with data preparation , lets separate the data now.We will filter it by train & test & remove data column from train data set & remove data & target variable Price from test data set.

In [57]:
h_train=h %>% filter(data=="train") %>% select(-data)
h_test=h %>% filter(data=="test") %>% select(-data,-Price)

**Regression Model Building with Random Forest- **

As we know that RandomForest has 4 parameters- mtry,ntree,maxnodes & nodessize.Let us set the parameter values that we want to try out.

In [60]:
param=list(mtry=c(5,10,20,30),
           ntree=c(50,100,200,500),
           maxnodes=c(5,10,15,20),
           nodesize=c(1,2,5,10))

The above leads to 444*4 i.e 256 possible combinations. We should technically CV performance of all these combination to find which one is the best. But that might be an overkill in terms of how much time it might take.If we are looking to do a 10 fold cross validation it will lead to 256X10 = 2560 randomforest models being built. If on an average each randomforest had 100 Trees, we are looking at 256,000 decision trees being built internally. This is beyond resource consuming. At the same time we need to realise that we dont really need to try out all the possible combination.Instead we can randomly select a much smaller subset to try.

Let's write a function which selects a random subset of combination.

In [62]:
subset_paras=function(full_list_para,n=10){
  all_comb=expand.grid(full_list_para)
  s=sample(1:nrow(all_comb),n)
  subset_para=all_comb[s,]
  return(subset_para)
}

If we pass the list of parameter values (full_list_para ) that we want to try and specify number of combinations (n) we want to try; it randomly selects those many combinations out of all possible parameter combinations and returns them in form of a dataframe.

We will take num_trials as 40 which is around 15% of total possible combinations (4^4 i.e 256). since a good value for num_trials is around 10-20% of total possible combination.

In [63]:
num_trials=40
my_params=subset_paras(param,num_trials)

We’ll be using cvTuning function from package cvTools to try out all these paramter combinations one by one. We’ll compare the cv error measure of all these and pick that as the best combination which results in lowest error.

In [65]:
library(cvTools)

Loading required package: lattice

Loading required package: robustbase



In [66]:
library(randomForest)

randomForest 4.6-14

Type rfNews() to see new features/changes/bug fixes.


Attaching package: ‘randomForest’


The following object is masked from ‘package:dplyr’:

    combine




In the below line of codes every time we find a parameter combination which has lowest error so far, it will be printed in the output.

In [67]:
myerror=9999999

for(i in 1:num_trials){
  print(paste0('starting iteration:',i))
 
  params=my_params[i,]
  k=cvTuning(randomForest,Price~.,
             data =h_train,
             tuning =params,
             folds = cvFolds(nrow(h_train), K=10, type = "random"),
             seed =2
  )
  score.this=k$cv[,2]
  if(score.this<myerror){
    print(params)
    
    myerror=score.this
    print(myerror)
    
    best_params=params
  }
   print('DONE')
  
}

[1] "starting iteration:1"
    mtry ntree maxnodes nodesize
153    5   200       10        5
[1] 510035.3
[1] "DONE"
[1] "starting iteration:2"
    mtry ntree maxnodes nodesize
207   20   500        5       10
[1] 493743.2
[1] "DONE"
[1] "starting iteration:3"
   mtry ntree maxnodes nodesize
32   30   500       10        1
[1] 438999.3
[1] "DONE"
[1] "starting iteration:4"
    mtry ntree maxnodes nodesize
240   30   500       15       10
[1] 414641.4
[1] "DONE"
[1] "starting iteration:5"
[1] "DONE"
[1] "starting iteration:6"
[1] "DONE"
[1] "starting iteration:7"
[1] "DONE"
[1] "starting iteration:8"
[1] "DONE"
[1] "starting iteration:9"
[1] "DONE"
[1] "starting iteration:10"
[1] "DONE"
[1] "starting iteration:11"
[1] "DONE"
[1] "starting iteration:12"
    mtry ntree maxnodes nodesize
248   30   100       20       10
[1] 404678.1
[1] "DONE"
[1] "starting iteration:13"
[1] "DONE"
[1] "starting iteration:14"
[1] "DONE"
[1] "starting iteration:15"
[1] "DONE"
[1] "starting iteration:16"
[1]

To know tentative performance measure , that’ll be latest value of myerror

In [68]:
myerror

[1] 404678.1

And Our best_params comes out to be

In [69]:
best_params

mtry ntree maxnodes nodesize
248 30   100   20       10

**Model Building- **

Now we have best values of paramteres from cross validation. We’ll use these values to build our RandomForest model on entire training data and use that for prediction on test.

In [70]:
h.rf.final=randomForest(Price~.,
                         mtry=best_params$mtry,
                         ntree=best_params$ntree,
                         maxnodes=best_params$maxnodes,
                         nodesize=best_params$nodesize,
                         data=h_train)

**Prediction on Test data set- **

In [72]:
test.pred=predict(h.rf.final,newdata = h_test)

In order to see first few records of predicted property Price for the test data set we need to use head function as under.

In [73]:
head(test.pred)

1         2         3         4         5         6 
 461177.7  616354.2  555588.1 1257527.5 1074786.6  952308.4